### PR TITLE
fix for 16-way pl310 

### PR DESCRIPTION
**Pull request by etienne-lms**
Get number of ways from the PL310 configuration AUX_CTRL[bit#16].

Can someone test on imx ?

### CODE REVIEW

**Comment by jbech-linaro** [comment]
@MrVan , see @etienne-lms comment above where he asks if someone using imx can test this. Would you mind?

**Comment by jbech-linaro** [comment]
@MrVan , I see, let's see if there is someone else watching this that can try it out. Thanks anyhow.

**Comment by jforissier** [comment]
Should this be rebased and merged maybe?

**Comment by etienne-lms** [comment]
I think this is needed but i dont' have hw to test against.
Merge and wait someone complains ? 
Close and wait for someone raises a issue?

File: core/arch/arm/kernel/tz_ssvce_pl310_a32.S [hunk 1]
```diff
@@ -134,8 +141,10 @@ FUNC arm_cl2_cleanbyway , :
 UNWIND(	.fnstart)
 
 	/* Clean by Way */
-	movw r1, #0x00FF
-	movt r1, #0x0000
+	mov r1, #0xFF
+	ldr r2, [r0, #PL310_AUX_CTRL]
+	tst r2, #PL310_AUX_16WAY_BIT
+	orrne r1, #0xFF00
```
Comment (jforissier) [inline core/arch/arm/kernel/tz_ssvce_pl310_a32.S]: This is used 3 times. Use a macro?

**Comment by jforissier** [comment]
@etienne-lms did you test on `plat-stm`? My understanding is that you did, but plat-stm have 8-way cache and you have no hardware with 16-way cache to test on. Is that correct?
If it is the case, then I think we can safely merge this. Because, in any case 16-way could not possible work before, right?

**Comment by etienne-lms** [comment]
@jforissier: yes, right, right and yes...

**Comment by MrVan** [comment]
sorry for this late test, I am just able to setup the environment on i.mx 6qsabresd board. The default code use 8 ways. Using this patch, imx linux kernel can not boot up. If using 8 ways, kernel can boot up.

**Comment by etienne-lms** [comment]
@MrVan, did you sabre fail to boot because this patch was applied ? 
I so, please confirm. This patch should not affect platforms using 8ways PL310.


Or did your sadre failed to boot when you apply the patch AND if you force AUX[bit#16]=1 ?
If so, that's normal. 

Actually, we are looking for a HW with a PL310 wired for 16-ways  associativity.

**Comment by MrVan** [comment]
@etienne-lms I am using i.mx 6qsabresd board.  These two options are defined.
```
$(call force,CFG_PL310,y)
$(call force,CFG_PL310_LOCKED,y)
```

If I remove the second line, without applying your patch, linux can not boot up.
If with the second line, and with your patch applied, linux can not boot up.

The log from Linux:

```
L2C-310 erratum 769419 enabled
L2C-310 cache controller enabled, 16 ways, 1024 kB
L2C-310: CACHE_ID 0x410000c7, AUX_CTRL 0x3c470800
```

Can bit16 be changed from 0 to 1? I am not sure about this. I did not touch this bit in my code.
So pl310 on i.MX6Q is 16 ways.

**Comment by jbech-linaro** [comment]
@MrVan I read this as the patch @etienne-lms proposes neither makes things better or worse on your setup, is that right?

**Comment by MrVan** [comment]
@jbech-linaro With this patch applied, kernel can not boot up.

I am not sure whether it is because this patch or something else about pl310. I enabled pl310 in kernel. And CFG_PL310_LOCKED in optee. But Without @etienne-lms patch, kernel is ok to boot up. After applied this patch, kernel can not boot up.

**Comment by etienne-lms** [comment]
thanks @MrVan. so this current patch corrupts your optee.

Note: you need to keep CFG_PL310_LOCKED=y, otherwise linux boots with some L2 cache enable whereas it shall boot cache off.

I wonder if the PL310 maintenance by ways (flush/inval/clean) are fully operational on your platform.
Do you exerience boots that sporadically fail without reason ? (i guess not). cache sync by ways is only used by optee at boot time.

Until this is clarified, we shall NOT merge this P-R.

**Comment by MrVan** [comment]
@etienne-lms There is a small bug in a9_plat_init.S about PL310 in plat-imx, I changed to use the following patch. I know that the original code use 16 ways, but I changed to 8 ways. This will be fixed later.

```
--- a/core/arch/arm/plat-imx/a9_plat_init.S
+++ b/core/arch/arm/plat-imx/a9_plat_init.S
@@ -56,6 +56,8 @@
 FUNC arm_cl2_config , :
 UNWIND(        .fnstart)

+       push {lr}
+
        /* disable PL310 ctrl */
        mov  r1, #0x0
        str  r1, [r0, #PL310_CTRL]
@@ -136,12 +138,9 @@ UNWIND(    .fnstart)
        str  r1, [r0, #PL310_POWER_CTRL]

        /* invalidate all cache ways */
+       bl arm_cl2_invbyway

-       movw r1, #0xFFFF
-       movt r1, #0x0000
-       str  r1, [r0, #PL310_INV_BY_WAY]
-
-       mov pc, lr
+       pop {pc}
```

After applying the upper patch, I did not met boot issues. After applying your patches, I use jtag to debug, and found that  when step execute `rfefd   sp!` in `smc_exit`, system hangs. If not apply your patch, step execute `rfefd sp!` and all is ok.

**Comment by MrVan** [comment]
@etienne-lms  From your patch:
```
 FUNC arm_cl2_lockallways , :
 UNWIND(        .fnstart)
-       add  r1, r0, #PL310_DCACHE_LOCKDOWN_BASE
-       mov r0, #PL310_NB_WAYS
+       add r1, r0, #PL310_DCACHE_LOCKDOWN_BASE
+       ldr r2, [r0, #PL310_AUX_CTRL]
+       tst r2, #PL310_AUX_16WAY_BIT
+       mov r0, #PL310_8_WAYS
+       addne r0, r0
```

Here r0 will be set to 0x10 on my platform. But from PL310 TRM, I found that there only `reg0-reg7`, so this piece code is wrong. With your patch applied, and revert the upper change to use 8, kernel boots up.

**Comment by MrVan** [comment]
@etienne-lms  According to your change, I changed to the following and kernel boot ok.
```
 FUNC arm_cl2_lockallways , :
 UNWIND(        .fnstart)
-       add  r1, r0, #PL310_DCACHE_LOCKDOWN_BASE
-       mov r0, #PL310_NB_WAYS
-       movw r2, #0xFFFF        /* LD ways constant */
+       add r1, r0, #PL310_DCACHE_LOCKDOWN_BASE
+       ldr r2, [r0, #PL310_AUX_CTRL]
+       tst r2, #PL310_AUX_16WAY_BIT
+       mov r0, #8              /* Reg 0 - Reg 7 */
+       mov r2, #0xFF           /* LD ways constant */
+       orrne r2, #0xFF00
```

**Comment by etienne-lms** [comment]
thanks @MrVan for the review and corrections. you are right. i will fix the patch and rebase.

**Comment by etienne-lms** [comment]
squashed, rebased.

**Comment by MrVan** [comment]
There is build break when only apply the first patch.

**Comment by etienne-lms** [comment]
true. change in "tz_ssvce_def.h" from the 2nd commit is required by the 1st.
thanks @MrVan, i will push back a fix.

**Comment by etienne-lms** [comment]
first commit fixed, rebased, + extra beauty in 2nd commit (magic value => #define)

**Comment by jforissier** [comment]
For commit "pl310: beautify":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by etienne-lms** [comment]
@MrVan, can you add your review tag (even tested-by tag if you have tested it) on this change so it can be merged?
useless, of course, you think this change (specifically the 1st commit "pl310: fix 16-way pl310 support") is not mature enough to be merged, in which case i will try to fix until we are all happy with it or we simply discard it.

**Comment by MrVan** [comment]
Reviewed-by: Peng Fan `<peng.fan@nxp.com>`
Tested-by: Peng Fan `<peng.fan@nxp.com>` on i.MX6Q-SDB

**Comment by etienne-lms** [comment]
thanks @MrVan, i'll rebase, and apply tags.

**Pull request merged**